tests/late_joiners.py

```diff
@@ -104,107 +107,106 @@ def run_requests(nodes, total_requests, start_id, final_msg, final_msg_id):
 
 
 def run(args):
-    hosts = ["localhost", "localhost", "localhost"]
+    hosts = ["localhost", "localhost", "localhost", "localhost"]
 
     with infra.ccf.network(
         hosts, args.binary_dir, args.debug_nodes, args.perf_nodes, pdb=args.pdb
     ) as network:
         network.start_and_join(args)
         first_node, backups = network.find_nodes()
         all_nodes = network.get_joined_nodes()
-
         term_info = find_primary(network)
         first_msg = "Hello, world!"
         second_msg = "Hello, world hello!"
+        catchup_msg = "Hey world!"
         final_msg = "Goodbye, world!"
 
-        nodes_to_kill = [backups[0]]
-        nodes_to_keep = [first_node, backups[1]]
-
-        # first timer determines after how many seconds each node will be suspended
-        if not args.skip_suspension:
-            timeouts = []
-            for node in all_nodes:
-                t = random.uniform(1, 10)
-                LOG.info(f"Initial timer for node {node.node_id} is {t} seconds...")
-                timeouts.append((t, node))
-
-            for t, node in timeouts:
-                tm = Timer(t, timeout, args=[node, True, args.election_timeout / 1000],)
-                tm.start()
-
-        LOG.info(
-            "Adding another node after f = 0 but before we need to send append entries"
-        )
-        # check that a new node can catch up naturally
-        new_node = network.create_and_trust_node(
-            lib_name=args.package, host="localhost", args=args,
-        )
-        assert new_node
-        nodes_to_keep.append(new_node)
-
         with first_node.node_client() as mc:
             check_commit = infra.checker.Checker(mc)
             check = infra.checker.Checker()
 
-            run_requests(all_nodes, TOTAL_REQUESTS, 0, first_msg, 1000)
+            run_requests(all_nodes, int(TOTAL_REQUESTS / 2), 0, first_msg, 1000)
             term_info.update(find_primary(network))
 
-            # check that new node has caught up ok
-            assert_node_up_to_date(check, new_node, first_msg, 1000)
-            # add new node to backups list
-            all_nodes.append(new_node)
+            nodes_to_kill = []
+            nodes_to_kill.append(backups[0])
+            nodes_to_keep = [first_node, backups[1], backups[2]]
 
             # check that a new node can catch up after all the requests
             LOG.info("Adding a very late joiner")
-            last_node = network.create_and_trust_node(
+            last_joiner_node = network.create_and_trust_node(
                 lib_name=args.package, host="localhost", args=args,
             )
-            assert last_node
-            nodes_to_keep.append(last_node)
+            assert last_joiner_node
+            nodes_to_keep.append(last_joiner_node)
 
-            run_requests(all_nodes, TOTAL_REQUESTS, 1001, second_msg, 2000)
+            # some requests to be processed simultaneously with the late joiner catching up
```

Comment:
I am not sure about the simultaneously here. It really depends on how long the new node takes to catch up. Is this something that this test relies on?

Reply:
It checks that the nodes can process requests and help a node catch up at the same time

Reply:
Yup. However, as I see it, by the time we get to this line, the new node might have already caught up. We'd have to suspend the new node as soon as it has joined to guarantee the "simultaneous" here.

Reply:
I can change the comment to not imply simultaneously so strongly if you think it is confusing

Reply:
Thanks. I just want to make sure that we know what this test tests and what it doesn't. If we want to test that a node can help another node catch up at the same time that it processed requests, we need another test. Otherwise, deleting the comment will do.

Reply:
This tests aims mostly to introduce as much randomness as it can while at the same time testing certain things. This helps discover a lot of consensus bugs. There isn't a plan on adding a specific test for what you are mentioning afaik.